### ISSUE TITLE
[Question] What does learning_starts do for DDPG, TD3 and SAC?

### ISSUE THREAD

**Issue by zhihanyang2022** [DLR-RM/stable-baselines3#458]
### Question

If time step is less than learning_starts, are exploration actions sampled from the action space or from the randomly initialized model (in the case of default settings)? 

Is the model updated during this period of time?

Thanks!

### Checklist

- [x] I have read the [documentation](https://stable-baselines3.readthedocs.io/en/master/) (**required**)
- [x] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo (**required**)

**Comment by pathway** [comment]
Hint: https://github.com/DLR-RM/stable-baselines3/blob/75b6f3b3b0f207456d9dcac2c6e86e8e2a22115f/stable_baselines3/common/off_policy_algorithm.py#L69

**Comment by araffin** [comment]
Hello,

>If time step is less than learning_starts, are exploration actions sampled from the action space or from the randomly initialized model (in the case of default settings)?
>Is the model updated during this period of time?

as @pathway you will find most of your answer in the docstrings.

To sum it up:
- `learning_starts` corresponds to the warmup phase, where actions are uniformly sampled to fill up the buffer before starting updating the network

The only special case is for gSDE for using RL on real robot (where uniform sampling is a bad idea), where the randomly initialized model + gSDE will be used to explore randomly and fill up the buffer.

**Comment by zhihanyang2022** [comment]
Thank you all for the very detailed reply. I understand the issue now.

In fact, the reason I had to ask this question is that I'm trying SB3 vs my own implementation on an environment that requires good exploration at the beginning to avoid local minima solution. SB3 and my own implementation performs very similarly on other tasks, but for this one task SB3 performs much better. This led me to question whether SB3 uses a different warmup strategy (and the answer is no with all your help).

Are there other ways SB3 **helps with exploration**? 

I have thought of some hypotheses of what SB3 might do but they don't seem true after I read the code. Examples include:

- Observation normalization (I've asked about this in another issue): no
- Different warmup style: no
- Policy initialization: e.g., in TD3, there seems to be no weight initialization

https://github.com/DLR-RM/stable-baselines3/blob/75b6f3b3b0f207456d9dcac2c6e86e8e2a22115f/stable_baselines3/td3/policies.py#L58

https://github.com/DLR-RM/stable-baselines3/blob/75b6f3b3b0f207456d9dcac2c6e86e8e2a22115f/stable_baselines3/common/torch_layers.py#L9

**Comment by araffin** [comment]
>Are there other ways SB3 helps with exploration?

We are following the original implementation. So nothing special.
As shown in our blog post, we benchmarked our codebase against reference implementation (the easiest to try is TD3).

>SB3 and my own implementation

which algorithm?

**Comment by zhihanyang2022** [comment]
This perhaps come at a surprise, but eventually me and my collaborator found that something that shouldn't be affecting DDPG's performance but does.

SpinUp's DDPG (and TD3 and SAC) uses

https://github.com/openai/spinningup/blob/master/spinup/algos/pytorch/ddpg/ddpg.py#L202

before it computes the actor loss and uses

https://github.com/openai/spinningup/blob/master/spinup/algos/pytorch/ddpg/ddpg.py#L212

after it computes the actor loss.

While this looks completely harmless (I also think it's harmless), we found that having vs not having these two lines gave us a big performance difference on an environment.

Without the two lines:

![image](https://user-images.githubusercontent.com/43589364/120737050-f4ffe080-c51f-11eb-9ebd-71d0b367d7bf.png)

With the two lines:

![image](https://user-images.githubusercontent.com/43589364/120737103-0ea12800-c520-11eb-839b-28187e647e19.png)

Note however that these two lines did not cause much trouble for other environments.

**Comment by araffin** [comment]
>While this looks completely harmless (I also think it's harmless), we found that having vs not having these two lines gave us a big performance difference on an environment.

Interesting... which algorithm was used here?

to be sure it has an impact, you can try to fix the random seed and activate/deactivate those gradient freezing and check that the results are the same (you don't need to do a full run for that).
Because the figure you show are still quite noisy, and I would try with even more seeds before drawing any conclusion.
